### PR TITLE
Writer space leak fix

### LINKED ISSUE #85: WriterC leaks space
`WriterC` is isomorphic to `WriterT` and thus leaks space: https://twitter.com/gabrielg439/status/659170544038707201

We should define it like `StateC` instead.

### PR DESCRIPTION
**Pull request by robrix**
This PR:

- [x] Adds `Functor`, `Applicative`, and `Monad` instances for `VoidC` and `WriterC`.
- [x] Adds benchmarks of `WriterC`/`VoidC` against their `Eff`-wrapped counterparts. (cf #89)
- [x] Fixes #85.

### CODE REVIEW

File: fused-effects.cabal [hunk 1]
```diff
@@ -55,6 +55,16 @@ library
...
 
+benchmark benchmark
+  type:               exitcode-stdio-1.0
+  main-is:            Bench.hs
+  build-depends:      base >=4.9 && <4.13
+                    , criterion
+                    , fused-effects
+  hs-source-dirs:     benchmark
+  default-language:   Haskell2010
+  ghc-options:        -threaded -rtsopts "-with-rtsopts=-N -A4m -n2m" -j
```
[the remark this answers is not in the document]
Comment (robrix) [reply fused-effects.cabal]: Unfortunately that causes `cabal check` to complain. Plus, I think it’s valuable to be able to run the benchmarks at different optimization levels and see how the performance varies.

**Comment by robrix** [comment]
@patrickt: Documented with a link to the mailing list post, and also noted the current limitations.

File: fused-effects.cabal (hunk 1 above)
[the remark this answers is not in the document]
Comment (patrickt) [reply fused-effects.cabal]: Weird re. `cabal check`, but good enough for me.

**Comment by patrickt** [comment]
Looks terrific.

**Pull request merged**